src/language-markdown/printer-markdown.js

```diff
@@ -557,66 +560,96 @@ function printTable(path, options, print) {
     contents[0].map(() => 3) // minimum width = 3 (---, :--, :-:, --:)
   );
 
-  return join(hardline, [
+  const compactTable = join(hardlineWithoutBreakParent, [
+    // compactTable is true
+    printRow(contents[0], true),
+    printSeparator(true),
+    join(
+      hardline,
+      contents.slice(1).map(rowContents => printRow(rowContents, true))
+    )
+  ]);
+
+  const alignedTable = join(hardlineWithoutBreakParent, [
     printRow(contents[0]),
     printSeparator(),
-    join(hardline, contents.slice(1).map(printRow))
+    join(
+      hardline,
+      contents.slice(1).map(rowContents => printRow(rowContents, false))
+    )
   ]);
 
-  function printSeparator() {
+  return group(ifBreak(compactTable, alignedTable));
+
+  function printSeparator(compactTable) {
     return concat([
       "| ",
       join(
         " | ",
         columnMaxWidths.map((width, index) => {
+          const spaces = compactTable ? 3 : width;
           switch (node.align[index]) {
             case "left":
-              return ":" + "-".repeat(width - 1);
+              return ":" + "-".repeat(spaces - 1);
             case "right":
-              return "-".repeat(width - 1) + ":";
+              return "-".repeat(spaces - 1) + ":";
             case "center":
-              return ":" + "-".repeat(width - 2) + ":";
+              return ":" + "-".repeat(spaces - 2) + ":";
             default:
-              return "-".repeat(width);
+              return "-".repeat(spaces);
           }
         })
       ),
       " |"
     ]);
   }
 
-  function printRow(rowContents) {
+  function printRow(rowContents, compactTable) {
     return concat([
       "| ",
       join(
         " | ",
         rowContents.map((rowContent, columnIndex) => {
           switch (node.align[columnIndex]) {
             case "right":
-              return alignRight(rowContent, columnMaxWidths[columnIndex]);
+              return alignRight(
+                rowContent,
+                columnMaxWidths[columnIndex],
+                compactTable
+              );
             case "center":
-              return alignCenter(rowContent, columnMaxWidths[columnIndex]);
+              return alignCenter(
+                rowContent,
+                columnMaxWidths[columnIndex],
+                compactTable
+              );
             default:
-              return alignLeft(rowContent, columnMaxWidths[columnIndex]);
+              return alignLeft(
+                rowContent,
+                columnMaxWidths[columnIndex],
+                compactTable
+              );
           }
         })
       ),
       " |"
     ]);
   }
 
-  function alignLeft(text, width) {
-    return concat([text, " ".repeat(width - privateUtil.getStringWidth(text))]);
+  function alignLeft(text, width, compactTable) {
```

Comment:
The changes to `alignLeft`/`alignRight`/`alignCenter` seems unnecessary, you can simply return `rowContent` in `rowContents.map` if `isCompact === true` to achieve this goal.